extensions/functions_arithmetic.yaml

```diff
@@ -362,3 +362,54 @@ aggregate_functions:
         decomposable: MANY
         intermediate: fp64?
         return: fp64?
+  - name: "std_dev"
+    description: Calculates standard-deviation for a set of values.
+    impls:
+      - args:
+          - options: [ SILENT, SATURATE, ERROR ]
+            required: false
+          - value: i8
+        nullability: DECLARED_OUTPUT
+        decomposable: MANY
+        intermediate: "fp64?"
+        return: i8?
+      - args:
+          - options: [ SILENT, SATURATE, ERROR ]
+            required: false
+          - value: i16
+        nullability: DECLARED_OUTPUT
+        decomposable: MANY
+        intermediate: "fp64?"
+        return: i16?
+      - args:
+          - options: [ SILENT, SATURATE, ERROR ]
+            required: false
+          - value: i32
+        nullability: DECLARED_OUTPUT
+        decomposable: MANY
+        intermediate: "fp64?"
+        return: i32?
+      - args:
+          - options: [ SILENT, SATURATE, ERROR ]
+            required: false
+          - value: i64
+        nullability: DECLARED_OUTPUT
+        decomposable: MANY
+        intermediate: "fp64?"
+        return: i64?
+      - args:
+          - options: [ SILENT, SATURATE, ERROR ]
+            required: false
+          - value: fp32
+        nullability: DECLARED_OUTPUT
+        decomposable: MANY
```

Comment:
I'm pretty sure standard deviation isn't decomposable. AFAIK it can only be computed once the mean is already known.

Reply:
Please correct me if I am wrong. 

Standard deviation is the `sort(variance)`. 

Variance can be calculated by using `mean` (`avg`) and deviation for each sample. 

So if we partition the data, we can calculate the mean for each partition in an embarassingly parallel manner. Then using an `allreduce` (aggregate in parallel) call we can find the global `mean`. Then we can calculate the `deviation` for each sample in parallel for each partition. Again we can use `allreduce` to sum these values up and also get the count of values in each partition. Now we have total sample count and deviation sum for all values. In the rank=0 process we can divide the sum by sample count. Then taking the `sqrt` we can calculate `std_dev`. 

Is this wrong? Am I not following the definition of decomposability?

Reply:
Even if this is correct, my definition is wrong, there has to be a `struct<V1, V2>`, I need to fix this either way.

Reply:
That works for parallelizing stddev in general, but violates causality for the way Substrait deals with decomposition. Decomposition in Substrait means separating the aggregation between independent plans, so for instance one plan is pushed to several storage nodes and does the first part of the aggregation, and then another plan operates on a column containing the results from those plans and finishes the computation. That's why the intermediate type matters for as far as Substrait is concerned; how it's parallelized at the level of a single plan is not Substrait's concern. When applied to stddev, you can first indeed compute the mean in a decomposable manner, but then you need that mean *and* all individual rows in order to compute the variance (which again you could do in parallel, but you don't have the individual rows anymore at this point). At best the intermediate type could be something like `struct<fp64, list<fp64>>` to track the mean and all the values, but that seems a bit silly; at that point you can make all aggregates "decomposable" by just making the intermediate type the list of all input values thus far.

Reply:
For reference, this is the way that Calcite does [decomposition of these functions](https://calcite.apache.org/javadocAggregate/org/apache/calcite/rel/rules/AggregateReduceFunctionsRule.html): 

```
AVG(x) → SUM(x) / COUNT(x)
STDDEV_POP(x) → SQRT( (SUM(x * x) - SUM(x) * SUM(x) / COUNT(x)) / COUNT(x))
STDDEV_SAMP(x) → SQRT( (SUM(x * x) - SUM(x) * SUM(x) / COUNT(x)) / CASE COUNT(x) WHEN 1 THEN NULL ELSE COUNT(x) - 1 END)
VAR_POP(x) → (SUM(x * x) - SUM(x) * SUM(x) / COUNT(x)) / COUNT(x)
VAR_SAMP(x) → (SUM(x * x) - SUM(x) * SUM(x) / COUNT(x)) / CASE COUNT(x) WHEN 1 THEN NULL ELSE COUNT(x) - 1 END
COVAR_POP(x, y) → (SUM(x * y) - SUM(x, y) * SUM(y, x) / REGR_COUNT(x, y)) / REGR_COUNT(x, y)
COVAR_SAMP(x, y) → (SUM(x * y) - SUM(x, y) * SUM(y, x) / REGR_COUNT(x, y)) / CASE REGR_COUNT(x, y) WHEN 1 THEN NULL ELSE REGR_COUNT(x, y) - 1 END
REGR_SXX(x, y) → REGR_COUNT(x, y) * VAR_POP(y)
REGR_SYY(x, y) → REGR_COUNT(x, y) * VAR_POP(x)
```

From there, SUM is decomposed into SUM0 in numerous places.

Reply:
Thanks for the detailed explanation @jacques-n 
Should these be included in an Appendix section somewhere in the documentation. For instance, there are terminologies/standards that the Substrait community is trying to bring forward. So it would be very helpful if pointers to such definitions are there with more facts.

Reply:
Who is the audience for the decomposition field?  Is this advice to consumers / planners on how they might be able to decompose something?  Is it a restriction that limits what decompositions are allowed?

Reply:
I think the audience is a hypothetical optimizer & task distribution engine. If it gets a plan with a decomposable aggregation function, it can use the decomposability information and intermediate type to split the plan up and distribute it over nodes. Using `[X]` to signify a column of type `X`, saying an aggregate function from `[X] -> Z` is `ONE`- or `MANY`-decomposable with intermediate type `Y` implies that `[X] -> Y` (`INITIAL_TO_INTERMEDIATE`) and `[Y] -> Z` (`INTERMEDIATE_TO_RESULT`) can also be used in plans, and that piping the results of a number of `[X] -> Y` invocations to a single `[Y] -> Z` invocation yields the same result as doing `[X] -> Z` on all `X`s in one go. If it's `MANY`-decomposable `[Y] -> Y` (`INTERMEDIATE_TO_INTERMEDIATE`) also exists, allowing a distributor to use more exotic distribution patterns as well. Selection between these modes of operation happen using the "aggregation phase" option when binding the aggregate function (the default being `INITIAL_TO_RESULT` for the `[X] -> Z` variant), so effectively you're declaring several aggregate functions in one go by defining one to be decomposable.

How an engine implements a function internally (parallelization strategy, distribution over nodes, etc.) is out of scope for Substrait. It's ultimately just an interface, after all; such things are implementation details that are intentionally hidden by that interface.